include/boost/asio/windows_sspi/stream.hpp

```diff
@@ -119,14 +162,67 @@ template <typename NextLayer, typename ConstBufferSequence> struct async_write_i
                                              net::buffer(m_message),
                                              net::transfer_exactly(m_message.size()),
                                              std::move(self));
-      self.complete(ec, net::buffer_size(m_buffer));
+      self.complete(ec, m_size_encrypted);
     }
   }
 
 private:
   NextLayer& m_next_layer;
   ConstBufferSequence m_buffer;
-  std::shared_ptr<sspi_impl> m_sspi;
+  std::shared_ptr<sspi_impl> m_sspi_impl;
+  std::vector<char> m_message;
+  size_t m_size_encrypted{0};
+};
+
+// TODO: Move away from this file
+template <typename NextLayer, typename MutableBufferSequence> struct async_read_impl : boost::asio::coroutine {
+  async_read_impl(NextLayer& next_layer, const MutableBufferSequence& buffer, std::shared_ptr<sspi_impl> sspi_impl)
+    : m_next_layer(next_layer)
+    , m_buffer(buffer)
+    , m_sspi_impl(std::move(sspi_impl)) {
+  }
+
+  template <typename Self> void operator()(Self& self, boost::system::error_code ec = {}, std::size_t length = 0) {
+    BOOST_ASIO_CORO_REENTER(*this) {
+      while(m_sspi_impl->decrypted_data.empty()) {
+        // TODO: Find some way to make the sspi_impl, the decrypt
+        // function or something else be responsible for keeping track
+        // of state and buffer(s)
+        m_message.resize(0x10000 - m_sspi_impl->encrypted_data.size());
+        BOOST_ASIO_CORO_YIELD net::async_read(m_next_layer,
+                                              net::buffer(m_message),
+                                              std::move(self));
+        if (ec && length == 0 && m_sspi_impl->decrypted_data.empty()) {
+          self.complete(ec, 0);
+          return;
+        }
+        m_message.resize(length);
+        auto sc = m_sspi_impl->decrypt(m_message);
+        m_message.clear();
+        if (sc == SEC_E_INCOMPLETE_MESSAGE) {
+          continue;
+        }
+        if (sc == SEC_I_CONTEXT_EXPIRED) {
+          self.complete(net::error::eof, 0);
```

Comment:
is this really EOF though?

Reply:
[`The DecryptMessage (Schannel) function returns SEC_I_CONTEXT_EXPIRED when the message sender has shut down the connection.`](https://docs.microsoft.com/en-us/windows/win32/secauthn/shutting-down-an-schannel-connection).

I guess that's best handled by returning EOF to the client? Is there some better error code to use?

The same documentation also states that this should be followed by shutting down the context (which has not yet been implemented). The openssl implementation requires the user to do that though, so it might at least be a good idea for me to look into this.

Reply:
I'm not sure what the right answer is. At some point you will have to define your own error codes, if anything to deliver "short read". Or I guess you could deliver the corresponding asio error codes.

Reply:
Looking more into it, I was wrong in that the openssl implementation requires the user to do anything.

The correct solution is to shut down the connection gracefully (by following the documentation from Microsoft) and *then* I think it would be OK to send EOF to the client.

The whole async shutdown thing needs to be implemented though, but in short, you are absolutely right that simply returning EOF here is not correct.